### DISCUSSION TITLE
actix-web-actors: impl Handler method how to obtain async  environment

### DISCUSSION THREAD

**Discussion by adminSxs** [actix/actix-web#2829] [Q&A]
example:

```rust
struct ws;

impl StreamHandler<Result<Message, ProtocolError>> for Ws {
    fn handle(&mut self, msg: Result<Message, ProtocolError>, ctx: &mut Self::Context) {
    
        }
 }

async fn handler1(){

}
```

I want to call handler1().await in handle method

**Comment by robjtede** [comment]
See this FAQ entry: https://github.com/actix/actix-web/wiki/FAQ#how-do-i-use-asyncawait-inside-a-websocket-handler

**Comment by Samielakkad** [comment]
`StreamHandler::handle` is synchronous because it runs inside the actor context. To run async work from there, turn the future into an actor future and spawn/wait it on the context.

Example:

```rust
use actix::{ActorFutureExt, ContextFutureSpawner, WrapFuture};

impl StreamHandler<Result<Message, ProtocolError>> for Ws {
    fn handle(&mut self, msg: Result<Message, ProtocolError>, ctx: &mut Self::Context) {
        async move {
            handler1().await
        }
        .into_actor(self)
        .map(|result, actor, ctx| {
            // use result here
            // ctx.text(...), actor.field = ..., etc.
        })
        .spawn(ctx);
    }
}
```

Use `.spawn(ctx)` if the actor can keep processing other messages while the async work runs.

Use `.wait(ctx)` if you need to pause the actor until the async work finishes:

```rust
future.into_actor(self).map(...).wait(ctx);
```

Be careful with `.wait(ctx)` for long operations because it can block the actor from handling pings/messages. For database/network calls, `.spawn(ctx)` is usually safer, then send the result back through the context when it completes.
